### ISSUE TITLE
[Mesa] Exception: "pipeline 1 does not meet strict OpenGL ES 3.1 requirements"

### ISSUE THREAD

**Issue by w-flo** [BearishSun/BansheeEngine#126]
Hi,

I use the Mesa driver on Ubuntu 17.10 for my Radeon HD 7870 (driver version from the Ubuntu -proposed repo because it fixes vulkan-smoketest). When I try to launch an example application after compiling in Debug mode, an exception is thrown because the OpenGL error/debug callback is called (severity medium, type PORTABILITY).

The error/debug message is "glValidateProgramPipeline: pipeline 1 does not meet strict OpenGL ES 3.1 requirements and may not be portable across desktop hardware". (I've checked mesa source code and noticed that this will only be reported in a debug context, which explains why it doesn't happen in Release mode.)

I've modified the debug callback to just log any calls unless the type is ERROR or UNDEFINED_BEHAVIOUR, and now the example application works like in Release mode (in my case, weird rendering) and the Mesa driver repeatedly logs debug messages about various pipelines not meeting strict requirements.

I've looked a little further into Mesa code to find out if it's a bug in Mesa or Banshee, and it seems like Mesa checks to make sure that input variables for each shader program match the set of output variables of the previous shader stage.

Since I don't know a lot about OpenGL, I'm still not sure if Mesa gets the check wrong, but looking into the OpenGL 4.5 spec, I see this:
```
With separable program objects, interfaces between shader stages may involve
the outputs from one program object and the inputs from a second program object.
For such interfaces, it is not possible to detect mismatches at link time, because the
programs are linked separately. When each such program is linked, all inputs or
outputs interfacing with another program stage are treated as active. The linker will
generate an executable that assumes the presence of a compatible program on the
other side of the interface. If a mismatch between programs occurs, no GL error is
generated, but some or all of the inputs on the interface will be undefined.
At an interface between program objects, the set of inputs and outputs are con-
sidered to match exactly if and only if:
• Every declared input block or variable must have a matching output, as de-
scribed above.
• There are no output blocks or user-defined output variables declared without
a matching input block or variable declaration.
```

The spec about matching variables ("described above"):
```
• An output variable is considered to match an input variable in the subsequent
shader if:
– the two variables match in name, type, and qualification, and neither
has a location qualifier, or
– the two variables are declared with the same location and
component layout qualifiers and match in type and qualification.
```

Some LOGDBG() calls in various places lead me to assume that this is used in one of the pipelines used by the example app as the vertex shader:
```
out vec2 xsv_TEXCOORD0;
```
But the fragment shader has this:
```
in vec2 bs_TEXCOORD0;
```

So this seems to be a mismatch as described by the OpenGL 4.5 spec, because they match in type and qualification, but not in name? Or maybe I got this wrong, in that case it might be a Mesa bug?

**Comment by BearishSun** [comment]
Sounds like that could be the problem. To fix the naming you probably just need to add a line of code in BsSLFXCompiler.cpp. Take a look at line 561:
`outputDesc.nameMangling.inputPrefix = "bs_";`

There is a similar option named `outputPrefix` which currently isn't being set. If you set that to "bs_" as well, the naming should hopefully be consistent and you can confirm if that is indeed the problem.

Note that after making the change, you'll have to force all shaders to rebuild. This can be done by deleting `Timestamp.asset` in both `Data\Engine` and `Data\Editor` folders, as well as `Data/Examples/Example.bsl.asset`. After doing that the next start of the engine will rebuild all the assets (might take a minute or few).

**Comment by w-flo** [comment]
Yep, that solves this portability warning and related exception in Debug mode! The ExampleGettingStarted now starts on Mesa (but I still have the check for GL_(EXT|ARB)_geometry_shader4 commented out, hoping that the geometry shaders in OpenGL Core are good enough…). Thanks :-) It doesn't solve my rendering issues though, so I'll keep looking, I don't think it's related to geometry shaders.

**Comment by BearishSun** [comment]
I've submitted a fix with ca96a44dd78edb10dbb16464940d8312af91c09f, thanks for confirming this was the issue.

In case you want to debug the rendering issue, I can come up with a few suggestions:
 - Try to get RenderDoc up and running. You can analyse every single OpenGL calls, including its inputs and outputs, and it should hopefully be enough to track down the problem without even going further down this list.
 - Camera has a RenderSettings property in which you can enable/disable a all kinds of rendering effects. You can tweak the example code in an attempt to disable most advanced rendering and try to narrow down the issue to a specific shader. After changing the settings make sure to call Camera->setRenderSettings so they actually get applied.
 - As a pure guess, I think the issue is likely due to variable offsets within the uniform buffer. There is code that generates uniform buffers based on the spec, which double-checking it matches what the shader has. It's possible that code might be slightly off.

**Comment by BearishSun** [comment]
Regarding the missing glMemoryBarrier issue you reported in chat, the solution is proposed as follows:
1. During rendering (more specifically, during parameter binding) check if load-store textures or writeable buffers are being written by the current shader. If so tag them as dirty.
2. During rendering check the dirty flags on any load-store textures and writeable buffers, and if they are present, execute the memory barrier before executing the draw/dispatch call.

In order to achieve step no.1 we need to know if shader performs an image or ssbo store. This can be achieved by extending XShaderCompiler so it outputs this information along with other information about shader parameters (currently it reports name, type and few other bits). I'm pretty sure at some point I added code for the compiler to track write-status of buffers/textures, for some other purpose, so hopefully it should just be the matter of exposing that information in the data reported after compilation (in ReflectionData struct).

For step no.2 something similar is needed, as memory barrier type depends on actual usage of a texture/buffer in the shader. So we need a way to check what that usage is. I believe the GLSL compiler in GLSLGpuProgram gives us enough information for that since GLSL uses different types for normal buffers/textures vs image buffers/textures, and this information is already queried and stored. Pretty sure we only need to care about GL_TEXTURE_FETCH_BARRIER_BIT (normal texture/buffer reads), GL_SHADER_IMAGE_ACCESS_BARRIER_BIT (image load/store) and GL_SHADER_STORAGE_BARRIER_BIT.

Reported data would then be passed along from shader compilation, through GpuParams to a call to GLRenderAPI::setGpuParams() which would then be able to handle both step no.1 and no.2 internally.

(Memory barrier GL_FRAMEBUFFER_BARRIER_BIT would also need to be executed when binding render targets that were previously being image stored to, but in this case it's easy to check the same dirty flag set by step no.1, in a call to GLRenderAPI::setRenderTarget).

P.S. Perhaps it's not a bad idea to add an render API call for advanced users to do it manually (with a flag to disable the automatic handling).
